compiler/rustc_hir/src/hir.rs

```diff
@@ -1319,6 +1328,16 @@ impl fmt::Display for AsyncGeneratorKind {
     }
 }
 
+impl AsyncGeneratorKind {
+    pub fn descr(&self) -> &'static str {
+        match self {
+            AsyncGeneratorKind::Block => "future (`async` block)",
+            AsyncGeneratorKind::Closure => "future (`async` closure body)",
+            AsyncGeneratorKind::Fn => "future (`async fn` body)",
```

Comment:
looking at the actual error message I think we should leave out the `future` and parentheses and just mention the `async` block and similar part. The fact that it's a future is technically not relevant for the type of the thing

Reply:
In my experience, people think of all these things as "futures", as that's the bound on every api that takes multiple futures of the same type (FuturesUnordered is how I hit this). Though I need to add a test where I try to use an async block and an async fn in the same api and see what it looks like

Reply:
hmm... once you have an `impl Future`, you will get errors about opaque types and not the ones you created. This is similar to how closures have a "closure type", but once you have them in an `impl Fn` you only get an error about opaque types.

As a more concrete example, look at the error message in this playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f1363fce8006b6776f59c0192ace4ddc

Talking about a "future" here is similar to talking about a "callable" in the case of closures. The actual type of the closure is an anonymous closure type.

Reply:
Hmm this is a good point. when I find time this weekend I'll add more tests. (Also the notes in those errors messages are nice, I should add something like that, and perhaps a note about these async generators being futures)